chore: updates to changelog before v1.0.0-rc.1 release

andynog opened the pull request:
Partially close: #3255

This PR updates the `CHANGELOG.md` file with the unreleased information updated before the `v1.0.0-rc.1` release cut (not that there's no `v1.0.0-rc.1` release in the changelog, this will happen only when we cut `v1.0.0` major release, no `unclog release` command was executed)

The changelog was updated using the command:
```
unclog build -a > CHANGELOG.md 
```

This PR also includes a change in the unclog `config.toml` to sort entries by `id` since this seem to be the old behavior and it's easier to diff the file with the historical `CHANGELOG.md` (otherwise it would be sorted alphabetically and this change in ordering for the past releases which is can be very confusing to diff and compare).

#### PR checklist

- [ ] Tests written/updated
- [X] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [X] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

melekes commented on .changelog/config.toml:
```diff
@@ -14,4 +14,4 @@ release_date_formats = [
 
 [change_set_sections]
-sort_entries_by = "entry-text"
+sort_entries_by = "id"
```
I'm curious why this was changed.

andynog replied:
I explained above, I think this change was added and the changelog was not re-generated after. It messes up with the current ordering, in the past I believe it was the id which makes more sense from a chronological perspective.

melekes reviewed (approved):
👍

cason reviewed (commented):
I don't have much background to understand this, but suggestions:

- Could we have a Table of Contents?
- Shouldn't updates be classified by "big" releases?
- Why unreleased stuff is on the changelog?

sergio-mena commented on CHANGELOG.md:
```diff
@@ -183,6 +165,33 @@
+  `timeout_prevote_delta` and `timeout_precommit_delta` into `timeout_round`
+  and `timeout_round_delta` accordingly
+  ([\#2895](https://github.com/cometbft/cometbft/pull/2895))
+`[abci/client]` Deprecate `SetResponseCallback(cb Callback)` in the `Client` interface as it is no
+longer used. ([\#3084](https://github.com/cometbft/cometbft/issues/3084))
+`[mempool]` Change the signature of `CheckTx` in the `Mempool` interface to
+`CheckTx(tx types.Tx, sender p2p.ID) (*abcicli.ReqRes, error)`.
+([\#1010](https://github.com/cometbft/cometbft/issues/1010), [\#3084](https://github.com/cometbft/cometbft/issues/3084))
```
The original file in `.changelog` dir has to be correctly formatted, and `unclog` called again.

andynog replied:
fixed

sergio-mena commented on CHANGELOG.md:
```diff
@@ -291,17 +320,34 @@
 - `[store]` When pruning force compaction of the database. ([\#1972](https://github.com/cometbft/cometbft/pull/1972))
 - `[metrics]` Added metrics to monitor state store access. ([\#1974](https://github.com/cometbft/cometbft/pull/1974))
 - `[metrics]` Added metrics to monitor block store access. ([\#1974](https://github.com/cometbft/cometbft/pull/1974))
+`[config]` Removed unused `[mempool.max_batch_bytes]` mempool parameter
+ ([\#2056](https://github.com/cometbft/cometbft/pull/2056/))
```
need to fix format in original `.changelog` file

andynog replied:
fixed

sergio-mena commented on CHANGELOG.md:
````diff
@@ -489,6 +598,56 @@ Also fixes a small bug in the mempool for an experimental feature.
+the change from v0.37.3 that bumped the minimum Go version to v1.21.
+
+### BUG FIXES
+
+- `[mempool]` Avoid infinite wait in transaction sending routine when
+  using experimental parameters to limiting transaction gossiping to peers
+  ([\#1654](https://github.com/cometbft/cometbft/pull/1654))
+
+### FEATURES
+
+- `[mempool]` Add `nop` mempool ([\#1643](https://github.com/cometbft/cometbft/pull/1643))
+
+  If you want to use it, change mempool's `type` to `nop`:
+
+  ```toml
+  [mempool]
+
+  # The type of mempool for this node to use.
+  #
+  # Possible types:
+  # - "flood" : concurrent linked list mempool with flooding gossip protocol
+  # (default)
+  # - "nop"   : nop-mempool (short for no operation; the ABCI app is responsible
+  # for storing, disseminating and proposing txs). "create_empty_blocks=false"
+  # is not supported.
+  type = "nop"
+  ```
+
+## v0.34.31
+
+*November 27, 2023*
+
+Fixes a small bug in the mempool for an experimental feature.
+
+### BUG FIXES
+
+- `[mempool]` Avoid infinite wait in transaction sending routine when
+  using experimental parameters to limiting transaction gossiping to peers
+  ([\#1654](https://github.com/cometbft/cometbft/pull/1654))
+
````
Why does this appear in `main`?

andynog replied:
will fix it as per our conversation

sergio-mena commented on CHANGELOG.md:
```diff
@@ -519,6 +678,78 @@ gossip.
   ([\#1558](https://github.com/cometbft/cometbft/pull/1558))
   ([\#1584](https://github.com/cometbft/cometbft/pull/1584))
 
+## v0.37.3
+
```
Same for this hunk, why are these changes here in `v1.x`?

andynog replied:
will fix it as per our conversation

andynog commented:
@sergio-mena I've added the changes requested. In the last commits I've added these changes:
- Re-added the `summary.md` doc.
- Removed releases that should not be in the `.changelog` folder. I've ensured that these entries matched the entries in their respective releases (folder diff) so we are not loosing anything.
- Many formatting fixes and adding information that was missing to ensure entries are consistent and accurate.
- Re-generated the `CHANGELOG.md` file after these changes were in place.

melekes commented on .changelog/unreleased/summary.md:
```diff
@@ -0,0 +1,48 @@
+*December 4, 2023*
```
```suggestion
*June 20, 2024*
```

melekes commented on CHANGELOG.md:
```diff
@@ -2,22 +2,196 @@
 
 ## Unreleased
 
+*December 4, 2023*
```
```suggestion
*June 20, 2024*
```

melekes reviewed (approved):
👍

sergio-mena commented on .changelog/unreleased/summary.md:
```diff
@@ -0,0 +1,48 @@
+   previously such changes required a new breaking release (which currently
+   takes 6 to 12 months to reach production use for many users). See [ADR
+   109](./docs/architecture/adr-109-reduce-go-api-surface.md) for more details.
+7. Proposer-Based Timestamps (PBTS) support. PBTS is a Byzantine fault-tolerant
+algorithm used by CometBFT for computing block times. See [spec](./spec/consensus/proposer-based-timestamp) doc for PBTS.
```
```suggestion
7. Proposer-Based Timestamps (PBTS) support. PBTS is a Byzantine fault-tolerant
    algorithm used by CometBFT for computing block times.
    When activated on a chain, it replaces the pre-existing BFT-time algorithm.
    See [spec](./spec/consensus/proposer-based-timestamp) doc for PBTS.
```

Thanks for adding this @andynog !

sergio-mena commented on .changelog/unreleased/summary.md:
```diff
@@ -0,0 +1,48 @@
+URLs. See the [upgrading guidelines](./UPGRADING.md) and specific changes below
+for more details.
+
+**NB: This version is still an alpha-series release, which means that
```
```suggestion
**NB: This version is still a release candidate, which means that
```

andynog replied:
good catch! Thanks !

sergio-mena commented on .changelog/unreleased/summary.md:
```diff
@@ -0,0 +1,48 @@
+for more details.
+
+**NB: This version is still an alpha-series release, which means that
+API-breaking changes might still be introduced until such time that a _release
```
```suggestion
API-breaking changes, although very unlikely, might still be introduced until such time that a _release
```

sergio-mena commented on .changelog/unreleased/summary.md:
```diff
@@ -0,0 +1,48 @@
+for more details.
+
+**NB: This version is still an alpha-series release, which means that
+API-breaking changes might still be introduced until such time that a _release
+candidate_ is cut.** See [RELEASES.md](./RELEASES.md) for more information on
```
```suggestion
API-breaking changes, although very unlikely, might still be introduced
before the final release.** See [RELEASES.md](./RELEASES.md) for more information on
```
